File: DynamicProgramming/Knapsack/knapsack.cpp
#include <stdio.h>
#include <algorithm>

using namespace std;

int maxSum(int*, int, int);
int getSum(int, int, int, int, int*);

int main(){
  int cases, n, target;
  scanf("%d", &cases);

  while(cases--){
    scanf("%d %d", &n, &target);
    int* array = new int[n];

    for(int i=0; i<n; i++){
      scanf("%d", &array[i]);
    }

    int sum = maxSum(array, target, n); 
    printf("%d\n", sum);
  }

  return 0;
}

int maxSum(int* array, int target, int size){
  return getSum(0, 0, target, size, array);
}

int getSum(int index, int sum, int target,int size, int* array){
  if(target == 0)
    return sum;

  if(index == size){
    return sum;
  }

  if(array[index] > target){
    return getSum(index+1, sum, target, size, array); 
  }

  return max(getSum(index, sum+array[index], target-array[index], size, array),
             getSum(index+1, sum, target, size, array));
}

File: DynamicProgramming/Knapsack/knapsack.ex
defmodule Knapsack do

  def read_cases_number() do
    {n, ""} = IO.gets("")
              |> String.replace("\n","")
              |> Integer.parse()
    n
  end
  
  def read_array() do
    IO.gets("")
    |> String.replace("\n","")
    |> String.split()
    |> Enum.map(fn(array_item) -> 
      {number, _} = Integer.parse(array_item)
      number
    end)
  end

  def get_sum([], sum, _) do
    sum
  end

  def get_sum(_, sum, 0) do
    sum 
  end

  def get_sum(list, sum, target) do
    [current_item|t] = list

    case current_item do
      current_item when current_item > target ->
        get_sum(t, sum, target)
      _ -> get_sum(list, sum+current_item, target-current_item)
           |> max(get_sum(t, sum, target))
    end 
  end

end

1..Knapsack.read_cases_number()
|> Enum.each(fn(_) -> 
  [_, target] = Knapsack.read_array()

  Knapsack.read_array()
  |> Knapsack.get_sum(0, target)
  |> IO.puts
end)

